src/sage/topology/moment_angle_complex.py

```diff
@@ -814,3 +819,712 @@ def has_trivial_lowest_deg_massey_product(self):
+        structure (multiplication on the right-hand side) is defined
+        in a certain way. See
+        :meth:`.moment_angle_complex.CohomologyRing.Element.cup_product`
+        for more information.
+
+    .. NOTE::
+
+        This is not intended to be created directly by the user, but instead
+        via the :meth:`.moment_angle_complex.MomentAngleComplex.cohomology_ring`
+        method.
+
+    INPUT:
+
+    - ``base_ring`` -- the base_ring over which we compute the cohomology
+    - ``moment_angle_complex`` -- the moment-angle complex whose homology
+      we are computing
+
+    EXAMPLES::
+
+        sage: Z = MomentAngleComplex([[1,2,3], [1,2,4], [3,5], [4,5]])
+        sage: H = Z.cohomology_ring(); H
+        Cohomology module of Moment-angle complex of Simplicial complex with
+        vertex set (1, 2, 3, 4, 5) and facets {(3, 5), (4, 5), (1, 2, 3),
+        (1, 2, 4)} over Rational Field
+        sage: a = H.an_element(); a
+        2*h^{0,0} + 2*h^{3,0} + 3*h^{3,1}
+        sage: b = H.basis()[3,2]
+
+    We can compute the cup product::
+
+        sage: a.cup_product(b)
+        2*h^{3,2} - 5*h^{6,0}
+        sage: a * a
+        4*h^{0,0} + 8*h^{3,0} + 12*h^{3,1}
+
+        sage: RP2 = simplicial_complexes.RealProjectivePlane()
+        sage: Z = MomentAngleComplex(RP2)
+        sage: H = Z.cohomology_ring(GF(2))
+        sage: x = H.basis(5)[5, 0]
+        sage: y = H.basis(5)[5, 1]
+        sage: x * y
+        0
+    """
+    def __init__(self, base_ring, moment_angle_complex):
+        """
+        Initialize ``self``.
+
+        TESTS::
+
+            sage: S2 = simplicial_complexes.Sphere(2)
+            sage: Z = MomentAngleComplex(S2)
+            sage: H = Z.cohomology_ring(GF(5))
+            sage: TestSuite(H).run()
+            sage: H = Z.cohomology_ring(ZZ)
+            sage: TestSuite(H).run()
+        """
+        self._complex = moment_angle_complex
+        self._base_ring = base_ring
+
+        vertices = moment_angle_complex._simplicial_complex.vertices()
+        n = len(vertices)
+        self._graded_indices = {}
+
+        # Will be used for storing information about the subcomplexes
+        # from which we compute the cohomology
+        self._gens = {}
+        indices = []
+        for deg in range(moment_angle_complex.dimension() + 1):
+            num_of_gens = 0
+            self._gens[deg] = []
+            for i in range(n+1):
+                for x in combinations(vertices, i):
+                    S = moment_angle_complex._simplicial_complex.generated_subcomplex(x, is_mutable=False)
+                    # Because of the empty combination
+                    if len(S.vertices()) > 0 and isinstance(S.cohomology(deg-i-1, base_ring, generators=True), list):
+                        chmlgy = S.cohomology(deg-i-1, base_ring, generators=True)
+                        for y in chmlgy:
+                            self._gens[deg].append((set(x), deg-i-1, y))
```

Comment:
```suggestion
                            self._gens[deg].append((frozenset(x), deg-i-1, y))
```
Better to not allow accidental mutations.